ui/tutorial-renderer/src/main/java/org/cru/godtools/tutorial/activity/TutorialActivity.kt

```diff
@@ -0,0 +1,156 @@
+
+    // region Lifecycle
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.tutorial_activity)
+        setUpToolbar()
+    }
+
+    override fun onContentChanged() {
+        super.onContentChanged()
+        setupViewPager()
+    }
+
+    override fun onStart() {
+        super.onStart()
+        pageSet.feature?.let { Settings.getInstance(this).setFeatureDiscovered(it) }
+    }
+
+    private var tutorialMenu: Menu? = null
+
+    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
+        pageSet.menu?.let {
+            menuInflater.inflate(it, menu)
+            tutorialMenu = menu
+            setMenuVisibility(false)
+            return super.onCreateOptionsMenu(menu)
+        }
+        setHomeLinkVisibility(true)
+        return super.onCreateOptionsMenu(menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
+        when (item?.itemId) {
+            R.id.onboarding_close_menu -> onCloseClicked()
+        }
+        return super.onOptionsItemSelected(item)
+    }
+
+    // endregion Lifecycle
+
+    // region ViewPager
+    private var viewPager: ViewPager? = null
+
+    private fun setupViewPager() {
+        viewPager = findViewById<ViewPager>(R.id.tutorial_viewpager)?.also {
+            it.adapter = TutorialPagerAdapter(pageSet.pages, this)
+            it.setupIndicator()
+        }
+    }
+
+    private fun ViewPager.setupIndicator() {
+        this@TutorialActivity.findViewById<CircleIndicator>(R.id.on_boarding_indicator)?.let { indicator ->
+            indicator.setViewPager(this)
+            addOnPageChangeListener(object : ViewPager.SimpleOnPageChangeListener() {
+                override fun onPageSelected(position: Int) {
+                    updateIndicatorVisibility(indicator, position)
+                    updateMenuVisibility(position)
+                }
+            })
+            updateIndicatorVisibility(indicator)
+            updateMenuVisibility()
+        }
+    }
+
+    private fun ViewPager.updateIndicatorVisibility(indicator: CircleIndicator, page: Int = currentItem) {
+        indicator.visibility = if (pageSet.pages[page].showIndicator) View.VISIBLE else View.GONE
+    }
+
+    private fun ViewPager.updateMenuVisibility(page: Int = currentItem) {
+        setMenuVisibility(pageSet.pages[page].showMenu)
+        setHomeLinkVisibility(pageSet.pages[page].showHomeLink)
+    }
+    // endregion ViewPager
+
+    // region ToolBar
+
+    private fun setUpToolbar() {
+        val toolbar: Toolbar = findViewById(R.id.tutorial_toolbar)
+        setSupportActionBar(toolbar)
+        supportActionBar?.setDisplayShowTitleEnabled(false)
+    }
+
+    private fun setMenuVisibility(isToolbarVisible: Boolean) {
+        var i = 0
+        val size = tutorialMenu?.size() ?: 0
+        while (i < size) {
+            tutorialMenu?.getItem(i)?.isVisible = isToolbarVisible
+            i++
+        }
```

Comment:
there should be a much simpler way to do this in Kotlin

Reply:
`implementation "androidx.core:core-ktx:${deps.androidX.core}"`
`tutorialMenu?.forEach { it.isVisible = isToolbarVisible }`

Reply:
@frett it looks like you did this one as well?